pkg/controllerutils/worker.go

```diff
@@ -1,128 +0,0 @@
-// Copyright (c) 2018 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
-//
-// Licensed under the Apache License, Version 2.0 (the "License");
-// you may not use this file except in compliance with the License.
-// You may obtain a copy of the License at
-//
-//      http://www.apache.org/licenses/LICENSE-2.0
-//
-// Unless required by applicable law or agreed to in writing, software
-// distributed under the License is distributed on an "AS IS" BASIS,
-// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-// See the License for the specific language governing permissions and
-// limitations under the License.
-
-package controllerutils
-
-import (
-	"context"
-	"fmt"
-	"sync"
-	"time"
-
-	"github.com/go-logr/logr"
-	"k8s.io/apimachinery/pkg/types"
-	"k8s.io/apimachinery/pkg/util/wait"
-	"k8s.io/client-go/tools/cache"
-	"k8s.io/client-go/util/workqueue"
-	logf "sigs.k8s.io/controller-runtime/pkg/log"
-	"sigs.k8s.io/controller-runtime/pkg/reconcile"
-)
-
-// WorkerOptions are options for a controller's worker.
-type WorkerOptions struct {
-	logger logr.Logger
-}
-
-// WorkerOption is a func that mutates WorkerOptions.
-type WorkerOption func(*WorkerOptions)
-
-// WithLogger configures the logr.Logger to use for a controller worker.
-// If set, a logger preconfigured with the name and namespace field will be injected into the context.Context on
-// each Reconcile call. It can be retrieved via log.FromContext.
-func WithLogger(logger logr.Logger) WorkerOption {
-	return func(options *WorkerOptions) {
-		options.logger = logger
-	}
-}
-
-// CreateWorker creates and runs a worker goroutine that just processes items in the
-// specified queue. The worker will run until ctx is cancelled. The worker will be
-// added to the wait group when started and marked done when finished.
-func CreateWorker(ctx context.Context, queue workqueue.RateLimitingInterface, resourceType string, reconciler reconcile.Reconciler, waitGroup *sync.WaitGroup, workerCh chan<- int, opts ...WorkerOption) {
```

Comment:
Recently we were discussing with folks that there are too many breaking changes wrt to golang API changes - especially in the extension library. Are we sure that no extension or other project is not using these func? Maybe we should first deprecate them and schedule removal in 2-3 releases.

Reply:
I don't know, I haven't checked, but it also shouldn't matter much given that updating the g/g dependency is a conscious action of the developer and does not happen automatically/without control. I don't see much benefit with artificially delaying such cleanups since such practice just shifts such required actions by 2-3 weeks globally. Eventually, users of this function have to adapt anyways, whether it's today or tomorrow.

Reply:
> I don't know, I haven't checked, but it also shouldn't matter much given that updating the g/g dependency is a conscious action of the developer and does not happen automatically/without control.

Yes, today it does not happen automatically but the goal is to have it automated or semi-automated. Ref https://github.com/gardener/gardener/issues/6328

I believe we should establish a rule like "make sure first to deprecate funcs and remove them in n releases".
In this case the breaking change is not that big and any potential consumer could adapt easily by duplicating the funcs on their side. But still, can I ask you to add a release not of type `breaking dependency` for the removed funcs and what is their replacement?

Reply:
Even if it's automated then this will not be merged automatically. PRs will be opened and then it's still a conscious decision of the developer to merge them.

I definitely agree with adding release notes for dependencies (that's why we have them - just added it here, thanks for the reminder).
However, I don't see any advantage about establish such rule for Golang dependencies. It just increases operational complexity and doesn't really solve any problem (it's not really "solvable", I believe).
For changes related to APIs used by users which don't have control about the rollout of the changes it's a completely different  story, of course.